roms_tools/utils.py

```diff
@@ -0,0 +1,169 @@
+                + [eta_coarse_domain_size] * (nx - 2)
+                + [eta_coarse_domain_size + n_eta_ghost_cells]
+            )
+
+    if ny > 1:
+        partitioned_sizes["xi_rho"] = (
+            [xi_rho_domain_size + n_xi_ghost_cells]
+            + [xi_rho_domain_size] * (ny - 2)
+            + [xi_rho_domain_size + n_xi_ghost_cells]
+        )
+
+        if "xi_coarse" in dims_to_partition:
+            partitioned_sizes["xi_coarse"] = (
+                [xi_coarse_domain_size + n_xi_ghost_cells]
+                + [xi_coarse_domain_size] * (ny - 2)
+                + [xi_coarse_domain_size + n_xi_ghost_cells]
+            )
+
+    def cumsum(pmf):
+        """Implementation of cumsum which ensures the result starts with zero"""
+        cdf = np.empty(len(pmf) + 1, dtype=int)
+        cdf[0] = 0
+        np.cumsum(pmf, out=cdf[1:])
+        return cdf
+
+    partitioned_datasets = []
+    for j in range(ny):
+        for i in range(nx):
+            # file_number = i + (j * ny)
+
+            eta_rho_partition_indices = cumsum(partitioned_sizes["eta_rho"])
+            xi_rho_partition_indices = cumsum(partitioned_sizes["xi_rho"])
+
+            indexers = {
+                "eta_rho": slice(
+                    int(eta_rho_partition_indices[i]),
+                    int(eta_rho_partition_indices[i + 1]),
+                ),
+                "xi_rho": slice(
+                    int(xi_rho_partition_indices[j]),
+                    int(xi_rho_partition_indices[j + 1]),
+                ),
+            }
+
+            if "eta_v" in dims_to_partition:
+                eta_v_partition_indices = cumsum(partitioned_sizes["eta_v"])
+                indexers["eta_v"] = slice(
+                    int(eta_v_partition_indices[i]),
+                    int(eta_v_partition_indices[i + 1]),
+                )
+            if "xi_u" in dims_to_partition:
+                xi_u_partition_indices = cumsum(partitioned_sizes["xi_u"])
+                indexers["xi_u"] = slice(
+                    int(xi_u_partition_indices[j]), int(xi_u_partition_indices[j + 1])
+                )
+
+            if "eta_coarse" in dims_to_partition:
+                eta_coarse_partition_indices = cumsum(partitioned_sizes["eta_coarse"])
+                indexers["eta_coarse"] = slice(
+                    int(eta_coarse_partition_indices[i]),
+                    int(eta_coarse_partition_indices[i + 1]),
+                )
+
+            if "xi_coarse" in dims_to_partition:
+                xi_coarse_partition_indices = cumsum(partitioned_sizes["xi_coarse"])
+                indexers["xi_coarse"] = slice(
+                    int(xi_coarse_partition_indices[j]),
+                    int(xi_coarse_partition_indices[j + 1]),
+                )
+
+            partitioned_ds = ds.isel(**indexers)
+
+            partitioned_datasets.append(partitioned_ds)
+
+    # TODO return the file number index too?
+    return partitioned_datasets
```

Comment:
The reason we might want to return the file number index is that `xr.save_mfdataset` requires a list of file names as well as a list of datasets.

Reply:
Now returns a list of file numbers too, see https://github.com/CWorthy-ocean/roms-tools/pull/101/commits/315cd3ef91eb36f158f9c4e702df73b25216df68